internal/table/table.go

```diff
@@ -0,0 +1,190 @@
+package table
+
+import (
+	"math"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/sliceutils"
+	"github.com/ActiveState/cli/internal/termutils"
+)
+
+const dash = "\u2500"
+const linebreak = "\n"
+const linebreakRune = '\n'
+const padding = 2
+
+type FormatFunc func(string, ...interface{}) string
+
+type row struct {
+	columns []string
+}
+
+type Table struct {
+	headers []string
+	rows    []row
+}
+
+func New(headers []string) *Table {
+	return &Table{headers, []row{}}
+}
+
+func (t *Table) AddRow(vs ...[]string) *Table {
+	for _, v := range vs {
+		t.rows = append(t.rows, row{v})
+	}
+	return t
+}
+
+func (t *Table) Render() string {
+	if len(t.rows) == 0 {
+		return ""
+	}
+
+	termWidth := termutils.GetWidth()
+	colWidths, total := t.calculateWidth(termWidth)
+
+	out := ""
+	out += "[NOTICE]" + renderRow(t.headers, colWidths) + "[/RESET]" + linebreak
+	out += "[DISABLED]" + strings.Repeat(dash, total) + "[/RESET]" + linebreak
+	for _, row := range t.rows {
+		out += renderRow(row.columns, colWidths) + linebreak
+	}
+
+	return strings.TrimRight(out, linebreak)
+}
+
+func (t *Table) calculateWidth(maxTotalWidth int) ([]int, int) {
+	// Calculate required width of each column
+	minWidth := padding * 2
+	colWidths := make([]int, len(t.headers))
+	columnTotal := 0
+	for n, header := range t.headers {
+		// Check header sizes
+		headerSize := utf8.RuneCountInString(header)
+		if currentSize, ok := sliceutils.GetInt(colWidths, n); !ok || currentSize < headerSize {
+			colWidths[n] = headerSize // Set width according to header size
+		}
+		// Check row column sizes
+		for _, row := range t.rows {
+			spanWidth := columnTotal
+			rowValueSize := 0
+			if rowValue, ok := sliceutils.GetString(row.columns, n); ok {
+				rowValueSize = utf8.RuneCountInString(rowValue)
+			}
+			if colWidths[n] < rowValueSize {
+				if len(row.columns) < len(t.headers) {
+					spanWidth += rowValueSize + (padding * 2) // This is a spanned column, so its width does not apply to the individual column
+				} else {
+					colWidths[n] = rowValueSize // Set width according to column size
+				}
+			}
+			if spanWidth > minWidth {
+				minWidth = spanWidth
+			}
+		}
+
+		// Add padding and update the total width so far
+		colWidths[n] += (padding * 2)
+		columnTotal += colWidths[n]
+	}
+
+	// Equalize widths by 20% of average width
+	// This is to prevent columns that are much larger than others from taking up most of the table width
```

Comment:
This function is starting to get pretty long and harder to grok. Is there any way we could split up some of this work?